Admin/BreadcrumbsBuilder.php

```diff
@@ -0,0 +1,143 @@
+        }
+
+        $menu = $admin->buildBreadcrumbs($action);
+
+        do {
+            $breadcrumbs[] = $menu;
+        } while ($menu = $menu->getParent());
+
+        $breadcrumbs = array_reverse($breadcrumbs);
+        array_shift($breadcrumbs);
+
+        return $breadcrumbs;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildBreadcrumbs(
+        AdminInterface $admin,
+        $action,
+        MenuItemInterface $menu = null
+    ) {
+        if (!$menu) {
+            $menu = $admin->getMenuFactory()->createItem('root');
+
+            $menu = $this->createMenuItem(
+                $admin,
+                $menu,
+                'dashboard',
+                'SonataAdminBundle',
+                array('uri' => $admin->getRouteGenerator()->generate(
+                    'sonata_admin_dashboard'
+                ))
+            );
+        }
+
+        $menu = $this->createMenuItem(
+            $admin,
+            $menu,
+            sprintf('%s_list', $admin->getClassnameLabel()),
+            null,
+            array(
+                'uri' => $admin->hasRoute('list') && $admin->isGranted('LIST') ?
+                $admin->generateUrl('list') :
+                null,
+            )
+        );
+
+        $childAdmin = $admin->getCurrentChildAdmin();
+
+        if ($childAdmin) {
+            $id = $admin->getRequest()->get($admin->getIdParameter());
+
+            $menu = $menu->addChild(
+                $admin->toString($admin->getSubject()),
+                array(
+                    'uri' => $admin->hasRoute('edit') && $admin->isGranted('EDIT') ?
+                    $admin->generateUrl('edit', array('id' => $id)) :
+                    null,
+                )
+            );
+
+            return $childAdmin->buildBreadcrumbs($action, $menu);
+        }
+
+        if ('list' === $action && $admin->isChild()) {
+            $menu->setUri(false);
+        } elseif ('create' !== $action && $admin->hasSubject()) {
+            $menu = $menu->addChild($admin->toString($admin->getSubject()));
+        } else {
+            $menu = $this->createMenuItem(
+                $admin,
+                $menu,
+                sprintf('%s_%s', $admin->getClassnameLabel(), $action)
+            );
+        }
+
+        return $menu;
+    }
+
+    /**
+     * Creates a new menu item from a simple name. The name is normalized and
+     * translated with the specified translation domain.
+     *
+     * @param AdminInterface    $admin             used for translation
+     * @param MenuItemInterface $menu              will be modified and returned
+     * @param string            $name              the source of the final label
+     * @param string            $translationDomain for label translation
+     * @param array             $options           menu item options
+     *
+     * @return MenuItemInterface
+     */
+    private function createMenuItem(
```

Comment:
> This is harder to read than having all argument on one line

This is only a question of habit. Remember when switching to PSR2 ? I bet you found things hard to read for one day or two, then everything go better. Plus, it makes it undeniably easier to edit. And things are actually harder to read when you have to scroll to read them.

> In a less personal argument, I think if signature are very long, that's mean the method maybe have to argument. Having multi-lines signature remove this impression and could open the door on too abusive code (not talking of your precisely, but globally.).

Yes, that's right, it's a bit less shocking. When the breadcrumbs builder becomes stateful, the admin and menu arguments will disapperar, and the method will be one-lined. But I bet you could also say : if there is a multiline signature, you immediately spot how many dependencies you have, even though they have long type hintings / variables names, so you get a **better** impression of your number of variables.

Looool for the Magento example, I hope we never get to that level ^^